### PR TITLE
Organized into cabal project and various other updates

### PR DESCRIPTION
**Pull request by jonesnoaht**
I organized the project into a cabal project, made some contributions to the Yocta script.

### CODE REVIEW

**Review by bjing — COMMENTED** [review]
I'm using these terms to indicate the importance of each comment: https://www.netlify.com/blog/2020/03/05/feedback-ladders-how-we-encode-code-reviews-at-netlify/

File: LICENSE [hunk 1]
```diff
@@ -1,21 +1,202 @@
...
-
-THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
-IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-SOFTWARE.
+
+                                 Apache License
```
Comment (bjing) [inline LICENSE:2]: [boulder] I'm not sure about the nuances between MIT and Apache, I think it warrants a discussion?

File: app/Main.hs~ [hunk 2]
```diff
@@ -0,0 +1,8 @@
+module Main where
```
Comment (bjing) [inline app/Main.hs~:1]: [pebble] duplicate file?

File: src/.#Yocto 10.hs [hunk 3]
```diff
@@ -0,0 +1 @@
+noahjones@Noah-MBP.local.73354
```
Comment (bjing) [inline src/.#Yocto 10.hs:1]: What are all these small .hs files?

**Comment by jonesnoaht** [comment]
I just realized it's on-chain already.

**Review by bjing — COMMENTED** [review]
[boulder] One thing I'm worried about Cabal. In production systems, is dependency hell. That's why at work we always favour `stack` because it's always easy to get reproducible builds.

However, I have not looked into Cabal in the past few years. Briefly checking its documentation doesn't give me the results I wanna see. I would like to know if Cabal has evolved to the point where it has solved library dependency issues as time goes on?

The good thing about `stack` is that each snapshot gives you a pretty good guarantee that the libraries work well with each other. If there's a newer library version needed, it can be imported explicitly with a version hash.

File: LICENSE (hunk 1 above)
Comment (jonesnoaht) [reply LICENSE:2]: I agree. I wasn't paying attention to the details that had been decided, and it looks like it's on-chain, so I closed the pull request to minimize the diffs.

File: src/.#Yocto 10.hs (hunk 3 above)
Comment (jonesnoaht) [reply src/.#Yocto 10.hs:1]: These are autosave files created by emacs. Emacs should only keep one autosave file, so it seems to have gone berserk when I ran something in cabal or stack.

File: app/Main.hs~ (hunk 2 above)
Comment (jonesnoaht) [reply app/Main.hs~:1]: I have to recursively apply `*~` to my `.gitignore` to keep emacs' backup files out of my commits. Evidently `*~` only worked at the top level.

**Review by jonesnoaht — COMMENTED** [review]
I closed the pull request because it's messy – there is no need to add this trash to the repo history – and because the contract is on-chain, so I would like to have a complete suggestion prior to suggesting an update.

File: app/Main.hs~ (hunk 2 above)
Comment (jonesnoaht) [reply app/Main.hs~:1]: I just realized that I forgot to escape the tilde. 🤦🏼‍♂️

**Comment by bjing** [comment]
> I closed the pull request because it's messy – there is no need to add this trash to the repo history – and because the contract is on-chain, so I would like to have a complete suggestion prior to suggesting an update.

I was going to have a discussion with @Riley-Kilgore about what kind of workflow we should have, like how we propose tasks, how things get reviewed, how to do testing etc.

When new devs join, I can see how this can get messy quickly. So I'd like to have the minimum amount of overhead while maintaining clarity in terms of how things are done. 

Can I get your email address or discord handle?

**Comment by jonesnoaht** [comment]
I'm running behind, but HMU. cloudy#2379

**Comment by jonesnoaht** [comment]
> [boulder] One thing I'm worried about Cabal. In production systems, is dependency hell. That's why at work we always favour `stack` because it's always easy to get reproducible builds.
> 
> 
> 
> However, I have not looked into Cabal in the past few years. Briefly checking its documentation doesn't give me the results I wanna see. I would like to know if Cabal has evolved to the point where it has solved library dependency issues as time goes on?
> 
> 
> 
> The good thing about `stack` is that each snapshot gives you a pretty good guarantee that the libraries work well with each other. If there's a newer library version needed, it can be imported explicitly with a version hash.

I have only worked with cabal and have only recently started working with stack. I understand that Stack makes dependency-management easier by bundling dependency versions iteratively. The philosophy in cabal is that you specify that your code is comparable with any version up to the dot update that would break the API.

**Comment by bjing** [comment]
> The philosophy in cabal is that you specify that your code is comparable with any version up to the dot update that would break the API.

Then you would have to worry about what versions of libs work with what in your Cabal configuration. 

The good thing about stack, is that you don't specify versions in your dependency declarations unless you have a need. Stack works out the versions for you depending on what snapshot you're on. Generally the only upgrade you worry about is the snapshot version which is directly tied to GHC version. For example: https://www.stackage.org/lts-18.17. 

Then if you have a need for a specific version of a library that's not in the latest snapshot you're using, as I said above, you can directly specify the version hash in stack.yaml. This minimises the amount of effort you spend on managing your dependencies which sometimes can be very annoying or boarderline frustrating.

**Pull request closed without merging**